### ISSUE TITLE
[Feature Request] Adjustable bloom lighting?

### ISSUE THREAD

**Issue by ViperAcidZX** [emoose/OutRun2006Tweaks#42]
I'm wondering could it be possible to implement to adjust the intensity of the bloom lighting with this plugin? I was re-watching Mr. Thunderwing's old comparison videos ([this one](https://www.youtube.com/watch?v=WwK2HlKXHJI&t=883s) for example) between the different versions of OutRun 2/C2C, one of the things he brought up in this comparison is the differences in bloom intensity between the Xbox, PC (vanilla), and the PS3's OutRun Online Arcade versions and the PS3 version notably has it much more vibrant than the prior ports (although YMMV in the intensity in some scenarios).

**Comment by brad86m** [comment]
Yeah, the PS3/360 version is a slightly trimmed port of the Lindbergh arcade version, but with the bloom cranked up further. Lindbergh still has a few graphical extras missing on PC. The big one would be the cool little sequences that happen at the end of a run. Like the space shuttle taking off, or the volcano erupting. These did make it back into the last PS2 Japanese Special Tours release, but not C2C.

**Comment by s-ilent** [comment]
As far as I can tell, the PC version doesn't have any bloom, only lens flare. The two effects are conflated together in most writeups about this on the net, but from poking around with RenderDoc there doesn't seem to be anything that looks like a bloom pass. If it's there, it probably needs to be turned on somehow. 

From looking at it with Xemu, he bloom effect on Xbox is very visible on the horizon where the clouds touch the buildings and totally absent on PC. There also seems to be some soft particle effect going on and an ambient shadow under the car. 
![xemu screenshot](https://github.com/emoose/OutRun2006Tweaks/assets/16026653/ec34ee89-52c5-4b88-955f-3fba73fb5a67)

**Comment by emoose** [comment]
Just noticed that there is a bunch of `DispGlowFilter` code still in C2C which I assume is the bloom effect, but it's behind a check for some byte near the Outrun2006.ini settings, nothing seems to set that byte, but if I patch it to 1 at startup... it gives a pretty ugly result:

![OR2006C2C_tqzjumQjDU](https://github.com/emoose/OutRun2006Tweaks/assets/1755499/d0ccd5c1-0730-4ce3-963a-5de518b73795)

Guess they probably intended to add some kind of "Enable Bloom" setting but couldn't get the bloom effect to render properly, so hid it behind that check & removed all the code that would set the "Enable Bloom" byte.

At least the code is mostly still included, not really sure where to start with trying to fix it up though, will try poking around with PIX.

(in case anyone wants to try it, here's a build that'll set the byte: 
[OutrunTweaks-DispGlowFilter.zip](https://github.com/user-attachments/files/16118918/OutrunTweaks-DispGlowFilter.zip)
Maybe it could just be some incompatibility with one of the other effects like fog/AA)

**Comment by lance8765** [comment]
"What if it snowed in Sunny Beach?" xD

**Comment by s-ilent** [comment]
Poking at it with Renderdoc, it seems like the bloom filtering is backwards. The scene being rendered gives the buildings/road/etc an alpha value of 1, and the sky an alpha of 0. It masks the bloom by the alpha channel, blacking out the sky, and then applies a few passes of blurring along with tweaking the colours before adding it back to the framebuffer and drawing the car and UI on top. 

Going by how it looks on Xbox, it should probably be blacking out everything but the sky. (Though there might be more to the effect that's missing.) It might be possible to fix by replacing or patching the relevant shader.

One other issue is that the bloom buffer is rendered at the same resolution as the scene. Ideally, it should be shrunk to some 480p-ish resolution at the screen aspect ratio before it's calculated and then scaled back up so the size of the glare is the same between resolutions.

**Comment by RIKLAVIEGASINTRODUCTION** [comment]
The color on the 'Legend' level seems to partially work. 
![ot2](https://github.com/emoose/OutRun2006Tweaks/assets/92535412/eea406b4-5c81-404e-91ec-a3e338389502)
It's not white at least.

**Comment by brad86m** [comment]
It was a shame that they toned it down, especially on the Xbox, as I'm sure they could have kept those gfx for the newer tracks, instead of bringing those downgrades over to the original levels too. Tyre smoke is also cut back on C2C Xbox when drifting.
Or, at least it looks more "clumpy", if that makes sense. Intro race start tyre spin smoke density seems the same though.

I'm not sure about flag man's stencil shadow on PC, though. Every other post has a soft shadow. I'm going to boot up my old 98 PC (cheated with an early dx9 card) later today and check the vanilla disc out and see if it is still the same.

**Comment by emoose** [comment]
Seems Online Arcade uses very similar `DispGlowFilter` code to the broken C2C stuff, a lot of it seems to match up but there are a couple changes made there, eg. I noticed it seems to setup 4 textures with the resolution halving for each one.
There's also some stuff about D3D tiling added to it which I think might be an X360 exclusive thing, so probably can't just copy the code directly from it...

Wonder if RenderDoc might show any difference with that running under Xenia, not sure if the DX12 that Xenia uses will be that comparable though.

**Comment by italospain** [comment]
Hi emoose

I am not able to get this effect working with the build you linked. (no change)
Did you try out to set the Resolution to 640x480 to see if the effect is resolution dependent. 
Thanks for your work !

**Comment by italospain** [comment]
Unbelievable !

The Option Fogging is actually the HDR Glow Effect that is not working.
You have to disable the option to get the right Colors and Contrasts like the Lindberg Version.

Btw I am the guy who found the lens flare file in the wrong folder.

I hope emoose can Restore the HDR Glow effect.

One step closer !
